### PR TITLE
LAB-748 Event tracking with Amplitude

### PR DESCRIPTION
**Pull request by pkuttappan**
**-------------Work Done-----------------------**

- Created analytics.ts file to initialize amplitude and create a common logEvent function.
- Added events based on taxonomy shared [in this ticket](https://coveord.atlassian.net/browse/LAB-748)

- Also the file creation was breaking locally and hence wasn't redirecting to create-file config. I simply added a check to see if the hostname is localhost and added filename in state.

**-------------Points to discuss------------------------**

- What should be format and information that needs to be passed when 'completed file edit' event is logged on file save?
- Should we rename home page in events logged to home screen?
- Should the amplitude key be added to state since we have 'Web Scraper Dev' project to track local development?

### CODE REVIEW

File: panel-app/src/components/analytics.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import * as amplitude from '@amplitude/analytics-browser';
+
+export function initializeAmplitude() {
+	let amplitudeKey = null;
+	try {
+		amplitudeKey = process.env.AMPLITUDE_KEY;
+		amplitude.init(amplitudeKey, { defaultTracking: false });
+	} catch (e) {}
+	console.log(amplitudeKey);
```
Comment (jdevost) [inline panel-app/src/components/analytics.ts:9]: Don't need to console.log the key anymore, it was just to make sure it was loaded properly

File: panel-app/src/components/create-config/create-config.tsx [hunk 2]
```diff
@@ -265,11 +276,12 @@ export class CreateConfig {
 
 	tabClicked(index: number) {
 		this.activeTab = index;
+		logEvent(`viewed file ${this.tabs[this.activeTab].toLowerCase()}`);
```
Comment (jdevost) [inline panel-app/src/components/create-config/create-config.tsx:279]: I don't think we need the `file` for the tabs.

**Comment by jdevost** [comment]
> * Should the amplitude key be added to state since we have 'Web Scraper Dev' project to track local development?

Not sure I get the question, but it looks like we don't need to add it to the state. 
I thought we need to do `if (key) amplitude.track(....)` to prevent errors if the key is undefined and amplitude is not initialized. Looks like it doesn't fail, so we are good the way you set it up.

**Comment by jdevost** [comment]
> Should we rename home page in events logged to home screen?

Why not, it sounds better. 
or even just `home`?

**Comment by jdevost** [comment]
For the save, we could just report on how many items there is in the json spec. 

For example, if I may:  (`json` is the spec) 

```js
// count how many exclude, metadata, subItems, etc. items are there
let counts = {};
json.forEach(i => {
  Object.keys(i).forEach(key => {
    let values = i[key];
    if (typeof values === 'object') {
      values = Object.values(values);
    }
    if (!(values instanceof Array)) {
      values = [values];
    }
    counts[key] = (counts[key] || 0) + (values || []).length;
  });
});
// for example: counts = {for: 2, exclude: 4, metadata: 3, subItems: 1, name: 1}

```

**Comment by pkuttappan** [comment]
> For the save, we could just report on how many items there is in the json spec.
> 
> For example, if I may: (`json` is the spec)
> 
> ```js
> // count how many exclude, metadata, subItems, etc. items are there
> let counts = {};
> json.forEach(i => {
>   Object.keys(i).forEach(key => {
>     let values = i[key];
>     if (typeof values === 'object') {
>       values = Object.values(values);
>     }
>     if (!(values instanceof Array)) {
>       values = [values];
>     }
>     counts[key] = (counts[key] || 0) + (values || []).length;
>   });
> });
> // for example: counts = {for: 2, exclude: 4, metadata: 3, subItems: 1, name: 1}
> ```

Makes sense. So right now we focus on getting just the count of the items? Do we want the selectors for exclude items?

**Comment by jdevost** [comment]
> > For the save, we could just report on how many items there is in the json spec.
> > For example, if I may: (`json` is the spec)
> > ```js
> > // count how many exclude, metadata, subItems, etc. items are there
> > let counts = {};
> > json.forEach(i => {
> >   Object.keys(i).forEach(key => {
> >     let values = i[key];
> >     if (typeof values === 'object') {
> >       values = Object.values(values);
> >     }
> >     if (!(values instanceof Array)) {
> >       values = [values];
> >     }
> >     counts[key] = (counts[key] || 0) + (values || []).length;
> >   });
> > });
> > // for example: counts = {for: 2, exclude: 4, metadata: 3, subItems: 1, name: 1}
> > ```
> 
> Makes sense. So right now we focus on getting just the count of the items? Do we want the selectors for exclude items?

Yes, you are right. How about? 
```js
// count how many exclude, metadata, subItems, etc. items are there
let payload = {};
json.forEach(i => {
  Object.keys(i).forEach(key => {
    let values = i[key];
    if (typeof values === 'object') {
      values = Object.values(values);
    }
    if (!(values instanceof Array)) {
      values = [values];
    }
    payload[key] = (payload[key] || 0) + (values || []).length;

    const excludeSelectors = payload.excludeSelectors || [];
    if (key === 'exclude') {
      values.forEach(v => {
        excludeSelectors.push(`[${v.type}] ${v.path}`);
      });
    }
    payload.excludeSelectors = excludeSelectors;
  });
});
// for example: 
// {
//   "for": 2,
//   "excludeSelectors": [
//     "[CSS] header, .header, *[role=\"header\"]",
//     "[CSS] footer, .footer",
//     "[CSS] noscript, nav",
//     "[CSS] script"
//   ],
//   "exclude": 4,
//   "metadata": 3,
//   "subItems": 1,
//   "name": 1
// }
```

**Comment by pkuttappan** [comment]
> > > For the save, we could just report on how many items there is in the json spec.
> > > For example, if I may: (`json` is the spec)
> > > ```js
> > > // count how many exclude, metadata, subItems, etc. items are there
> > > let counts = {};
> > > json.forEach(i => {
> > >   Object.keys(i).forEach(key => {
> > >     let values = i[key];
> > >     if (typeof values === 'object') {
> > >       values = Object.values(values);
> > >     }
> > >     if (!(values instanceof Array)) {
> > >       values = [values];
> > >     }
> > >     counts[key] = (counts[key] || 0) + (values || []).length;
> > >   });
> > > });
> > > // for example: counts = {for: 2, exclude: 4, metadata: 3, subItems: 1, name: 1}
> > > ```
> > 
> > 
> > Makes sense. So right now we focus on getting just the count of the items? Do we want the selectors for exclude items?
> 
> Yes, you are right. How about?
> 
> ```js
> // count how many exclude, metadata, subItems, etc. items are there
> let payload = {};
> json.forEach(i => {
>   Object.keys(i).forEach(key => {
>     let values = i[key];
>     if (typeof values === 'object') {
>       values = Object.values(values);
>     }
>     if (!(values instanceof Array)) {
>       values = [values];
>     }
>     payload[key] = (payload[key] || 0) + (values || []).length;
> 
>     const excludeSelectors = payload.excludeSelectors || [];
>     if (key === 'exclude') {
>       values.forEach(v => {
>         excludeSelectors.push(`[${v.type}] ${v.path}`);
>       });
>     }
>     payload.excludeSelectors = excludeSelectors;
>   });
> });
> // for example: 
> // {
> //   "for": 2,
> //   "excludeSelectors": [
> //     "[CSS] header, .header, *[role=\"header\"]",
> //     "[CSS] footer, .footer",
> //     "[CSS] noscript, nav",
> //     "[CSS] script"
> //   ],
> //   "exclude": 4,
> //   "metadata": 3,
> //   "subItems": 1,
> //   "name": 1
> // }
> ```

This is a wonderful suggestion. Includes all the data that we require. Thanks a lot.

**Comment by pkuttappan** [comment]
I added a pointer in the ReadMe about amplitude being used and also added a sheet under docs folder which includes the taxonomy we used.

**Review by jdevost — APPROVED** [review]

**Pull request merged**